### PR TITLE
WithDropSecurityCapabilities method added

### PR DESCRIPTION
**Pull request by yprokule**
Besides adding _SecurityCapability_ there're use cases to specify _Dropped_ capabilities.

### CODE REVIEW

File: pkg/pod/container.go [hunk 1]
```diff
@@ -113,6 +113,91 @@ func (builder *ContainerBuilder) WithSecurityCapabilities(sCapabilities []string
...
+		sCapabilitiesList = append(sCapabilitiesList, v1.Capability(capability))
+	}
+
+	// filter possible duplicated capabilities from user's input
+	sCapabilitiesList = uniqueCaps(sCapabilitiesList)
+
+	// filter conflicting caps between ADD and DROP
+	if builder.definition.SecurityContext != nil {
+		if builder.definition.SecurityContext.Capabilities != nil {
+			if builder.definition.SecurityContext.Capabilities.Add != nil {
```
Comment (trewest) [inline pkg/pod/container.go:141]: Can be simplified to one if statement:
```suggestion
        if builder.definition.SecurityContext != nil &&
		builder.definition.SecurityContext.Capabilities != nil &&
		builder.definition.SecurityContext.Capabilities.Add != nil {
```

File: pkg/pod/container.go [hunk 2]
```diff
@@ -113,6 +113,91 @@ func (builder *ContainerBuilder) WithSecurityCapabilities(sCapabilities []string
...
+			}
+		}
+	} else {
+		glog.V(100).Infof("Defining new SecurityContext with Capabilities.Drop")
+		builder.definition.SecurityContext = &v1.SecurityContext{
+			Capabilities: &v1.Capabilities{
+				Drop: sCapabilitiesList,
+			},
+		}
+	}
```
Comment (trewest) [inline pkg/pod/container.go:196]: Nesting conditional this far is a bit hard to follow. Would something like this work for you?
```
	if builder.definition.SecurityContext == nil {
		builder.definition.SecurityContext = new(v1.SecurityContext)
	}

	if builder.definition.SecurityContext.Capabilities == nil {
		builder.definition.SecurityContext.Capabilities = new(v1.Capabilities)
	}

	if !redefine {
		sCapabilitiesList = capsDifference(
			builder.definition.SecurityContext.Capabilities.Drop,
			sCapabilitiesList, true)

		glog.V(100).Infof("Updating existing SecurityContext.Capabilities.Drop list")
		builder.definition.SecurityContext.Capabilities.Drop = append(
			builder.definition.SecurityContext.Capabilities.Drop,
			sCapabilitiesList...)
	} else {
		glog.V(100).Infof("Resetting existing SecurityContext.Capabilities.Drop list")
		builder.definition.SecurityContext.Capabilities.Drop = sCapabilitiesList
	}
```

File: pkg/pod/container.go (hunk 1 above)
Comment (yprokule) [reply pkg/pod/container.go:141]: that makes sense, thanks

File: pkg/pod/container.go (hunk 2 above)
Comment (yprokule) [reply pkg/pod/container.go]: thanks, updated

File: pkg/pod/container.go [hunk 3]
```diff
@@ -113,6 +113,81 @@ func (builder *ContainerBuilder) WithSecurityCapabilities(sCapabilities []string
...
+// WithDropSecurityCapabilities drops SecurityCapabilities from the container definition.
+func (builder *ContainerBuilder) WithDropSecurityCapabilities(sCapabilities []string, redefine bool) *ContainerBuilder {
+	glog.V(100).Infof("Dropping a list of SecurityCapabilities %v from container %s",
+		sCapabilities, builder.definition.Name)
+
+	if !areCapabilitiesValid(sCapabilities) {
+		glog.V(100).Infof("Given SecurityCapabilities %v are not valid. Valid list %s",
+			sCapabilities, AllowedSCList)
+
+		builder.errorMsg = "one of the give securityCapabilities is invalid. Please extend allowed list or fix parameter"
```
Comment (achuzhoy) [inline pkg/pod/container.go:125]: 's/give/provided/'

"the allowed list"

File: pkg/pod/container.go [hunk 4]
```diff
@@ -113,6 +113,81 @@ func (builder *ContainerBuilder) WithSecurityCapabilities(sCapabilities []string
...
+			builder.errorMsg = "Conflicting ADD and DROP SecurityCapabilities"
+		}
+	}
+
+	if builder.errorMsg != "" {
+		return builder
+	}
+
+	if builder.definition.SecurityContext == nil {
+		glog.V(100).Infof("SecurityContext is nil initializing one")
```
Comment (achuzhoy) [inline pkg/pod/container.go:163]: Missing a period after nil.

File: pkg/pod/container.go [hunk 5]
```diff
@@ -113,6 +113,81 @@ func (builder *ContainerBuilder) WithSecurityCapabilities(sCapabilities []string
...
+
+		builder.definition.SecurityContext = new(v1.SecurityContext)
+	}
+
+	if builder.definition.SecurityContext.Capabilities == nil {
+		glog.V(100).Infof("Capabilities are nil initializing one")
```
Comment (achuzhoy) [inline pkg/pod/container.go:169]: These seem like 2 sentences (or am I wrong). Missing the period after nil.

File: pkg/pod/container.go [hunk 6]
```diff
@@ -259,3 +334,46 @@ func areCapabilitiesValid(capabilities []string) bool {
 
 	return valid
 }
+
+func uniqueCaps(sCap []v1.Capability) []v1.Capability {
```
Comment (achuzhoy) [inline pkg/pod/container.go:338]: missing a description for the function

File: pkg/pod/container.go [hunk 7]
```diff
@@ -259,3 +334,46 @@ func areCapabilitiesValid(capabilities []string) bool {
...
+	}
+
+	for key := range uniqMap {
+		uniqSlice = append(uniqSlice, key)
+	}
+
+	return uniqSlice
+}
+
+func capsDifference(aCaps, bCaps []v1.Capability, inversion bool) []v1.Capability {
```
Comment (achuzhoy) [inline pkg/pod/container.go:353]: missing a description for the function

File: pkg/pod/container.go (hunk 3 above)
Comment (yprokule) [reply pkg/pod/container.go:125]: thanks, updated

File: pkg/pod/container.go (hunk 4 above)
Comment (yprokule) [reply pkg/pod/container.go:163]: thanks, updated

File: pkg/pod/container.go (hunk 5 above)
Comment (yprokule) [reply pkg/pod/container.go:169]: thanks, updated

File: pkg/pod/container.go (hunk 6 above)
Comment (yprokule) [reply pkg/pod/container.go:338]: yup, just saw other functions without description so _followed  the pattern_ 😄  thanks, updated

File: pkg/pod/container.go (hunk 7 above)
Comment (yprokule) [reply pkg/pod/container.go:357]: thanks, updated

File: pkg/pod/container.go [hunk 8]
```diff
@@ -113,6 +113,82 @@ func (builder *ContainerBuilder) WithSecurityCapabilities(sCapabilities []string
...
+			sCapabilitiesList, false)
+
+		if len(confCapabilitiesList) > 0 {
+			glog.V(100).Infof("Conflicting ADD and DROP capabilities")
+
+			for _, mcap := range confCapabilitiesList {
+				glog.V(100).Infof("SecurityCapability %q already present in the Capabilities.Add list", mcap)
+			}
+
+			builder.errorMsg = "Conflicting ADD and DROP SecurityCapabilities"
```
Comment (cdvultur) [inline pkg/pod/container.go:155]: I would combine here the below `return builder` to spare an extra `if builder.errorMsg != ""`

**Review by cdvultur — CHANGES_REQUESTED** [review]
Added some comments

File: pkg/pod/container.go [hunk 9]
```diff
@@ -113,6 +113,82 @@ func (builder *ContainerBuilder) WithSecurityCapabilities(sCapabilities []string
...
+	var sCapabilitiesList []v1.Capability
+	for _, capability := range sCapabilities {
+		sCapabilitiesList = append(sCapabilitiesList, v1.Capability(capability))
+	}
+
+	// filter possible duplicated capabilities from user's input
+	sCapabilitiesList = uniqueCaps(sCapabilitiesList)
+	glog.V(100).Infof("Filtered user input: %v", sCapabilitiesList)
+
+	// filter conflicting caps between ADD and DROP
```
[the remark this answers is not in the document]
Comment (yprokule) [reply pkg/pod/container.go:138]: thanks, updated

File: pkg/pod/container.go (hunk 8 above)
Comment (yprokule) [reply pkg/pod/container.go:155]: thanks, updated

File: pkg/pod/container.go [hunk 10]
```diff
@@ -113,6 +113,82 @@ func (builder *ContainerBuilder) WithSecurityCapabilities(sCapabilities []string
...
+	}
+
+	if builder.definition.SecurityContext.Capabilities == nil {
+		glog.V(100).Infof("Capabilities are nil. Initializing one")
+
+		builder.definition.SecurityContext.Capabilities = new(v1.Capabilities)
+	}
+
+	if !redefine {
+		glog.V(100).Infof("Do not redefine SecurityContext.Capabilities")
```
[the remark this answers is not in the document]
Comment (yprokule) [reply pkg/pod/container.go:176]: thanks, updated

File: pkg/pod/container.go [hunk 11]
```diff
@@ -259,3 +335,49 @@ func areCapabilitiesValid(capabilities []string) bool {
...
+
+	for key := range uniqMap {
+		uniqSlice = append(uniqSlice, key)
+	}
+
+	return uniqSlice
+}
+
+// capsDifference based on the value of inversion parameter
+// returns an intersaction or difference of 2 Security Capabilities lists.
```
[the remark this answers is not in the document]
Comment (yprokule) [reply pkg/pod/container.go:356]: thanks, updated

File: pkg/pod/container.go [hunk 12]
```diff
@@ -113,6 +113,82 @@ func (builder *ContainerBuilder) WithSecurityCapabilities(sCapabilities []string
...
+	glog.V(100).Infof("Dropping a list of SecurityCapabilities %v from container %s",
+		sCapabilities, builder.definition.Name)
+
+	if !areCapabilitiesValid(sCapabilities) {
+		glog.V(100).Infof("Given SecurityCapabilities %v are not valid. Valid list %s",
+			sCapabilities, AllowedSCList)
+
+		builder.errorMsg = "one of the provided securityCapabilities is invalid. " +
+			"Please extend the allowed list or fix parameter"
+	}
```
[the remark this answers is not in the document]
Comment (yprokule) [reply pkg/pod/container.go:129]: done, thanks

File: pkg/pod/container.go [hunk 13]
```diff
@@ -113,6 +113,82 @@ func (builder *ContainerBuilder) WithSecurityCapabilities(sCapabilities []string
...
+		builder.definition.SecurityContext.Capabilities = new(v1.Capabilities)
+	}
+
+	if !redefine {
+		glog.V(100).Infof("Do not redefine SecurityContext.Capabilities")
+		glog.V(100).Infof("Filtering duplicated DROP capabilities - %v", sCapabilitiesList)
+		sCapabilitiesList = capsDifference(builder.definition.SecurityContext.Capabilities.Drop,
+			sCapabilitiesList, true)
+
+		glog.V(100).Infof("Updating existing SecurityContext.Capabilities.Drop list")
```
[the remark this answers is not in the document]
Comment (yprokule) [reply pkg/pod/container.go]: done, thanks

File: pkg/pod/container.go [hunk 14]
```diff
@@ -113,6 +113,82 @@ func (builder *ContainerBuilder) WithSecurityCapabilities(sCapabilities []string
...
+		builder.definition.SecurityContext.Capabilities.Drop = append(builder.definition.SecurityContext.Capabilities.Drop,
+			sCapabilitiesList...)
+	} else {
+		glog.V(100).Infof("Resetting existing SecurityContext.Capabilities.Drop list")
```
[the remark this answers is not in the document]
Comment (yprokule) [reply pkg/pod/container.go]: updated, thanks

File: pkg/pod/container.go [hunk 15]
```diff
@@ -259,3 +335,49 @@ func areCapabilitiesValid(capabilities []string) bool {
...
+	for key := range uniqMap {
+		uniqSlice = append(uniqSlice, key)
+	}
+
+	return uniqSlice
+}
+
+// capsDifference based on the value of inversion parameter
+// returns an intersaction or difference of 2 Security Capabilities lists.
+func capsDifference(aCaps, bCaps []v1.Capability, inversion bool) []v1.Capability {
```
[the remark this answers is not in the document]
Comment (yprokule) [reply pkg/pod/container.go]: works for me :) updated, thanks

File: pkg/pod/container.go [hunk 16]
```diff
@@ -259,3 +335,49 @@ func areCapabilitiesValid(capabilities []string) bool {
 
 	return valid
 }
+
+// uniqueCaps filters duplicated Security Capabilities from the provided list.
+func uniqueCaps(sCap []v1.Capability) []v1.Capability {
```
[the remark this answers is not in the document]
Comment (yprokule) [reply pkg/pod/container.go]: renamed, thanks

**Review by cdvultur — APPROVED** [review]
/lgtm

File: pkg/pod/container.go [hunk 17]
```diff
@@ -259,3 +335,65 @@ func areCapabilitiesValid(capabilities []string) bool {
...
+
+		if found {
+			resultCaps = append(resultCaps, bVal)
+		}
+	}
+
+	return resultCaps
+}
+
+// capabilitiesIntersection returns a difference of 2 Security Capabilities lists.
```
Comment (achuzhoy) [inline pkg/pod/container.go:378]: capabilitiesDifference returns...

Comment (yprokule) [reply pkg/pod/container.go]: updated, thanks
